packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,121 @@
+    // row.
+    @attr({ attribute: 'parent-selects-children'})
+    public parentSelectsChildren?: boolean;
+}
+```
+
+The `Table` will also provide a `row-expand-toggle` event for when a row is expanded/collapsed that will provide details to the client including which row was expanded/collapsed, and what its new state is:
+
+```ts
+interface TableRowExpandedEventDetail {
+    newState: boolean;
+    recordId?: string;
+}
+```
+
+Note: This event will _not_ be emitted for group rows.
+
+### Lazy Loading
+
+The APIs noted above will enable the client to lazy load data into the `Table`. This will essentially be accomplished with the following steps:
+
+1. Providing a field in a record of the table data that indicates whether that row of data is intended to be a parent row. Records that are intended to be parents must set the value of the field, whose name is specified by the `hasChildrenFieldName` attribute, to `true`.
+2. After providing the current data to the `Table` via the `setData` method, all rows that have a value of `true` in the field specified by the `hasChildrenFieldName` attribute will display and expand/collapse button.
+3. Users must register a handler for the `row-expand-toggle` event on the `Table` instance, and will receive that event upon clicking the expand/collapse button.
+4. The details of the handled event will include the id for the row that was expanded.
+5. The user must then create a set of data with rows where they provide a field with the name specified by the `parendIdFieldName` attribute that has the value of the `recordId` value supplied in the event details.
+6. The user then sets the data on the `Table` again with the `setData` method.
+
+Notes:
+
+-   It is up to the user to manage whether or not the children of a row has already been loaded in order to avoid recreation of their data and calling `setData` on the `Table` unnecessarily.
+
+### Translating flat list to Tanstack-understandable hierarchy
+
+Tanstack provides APIs for us to implement that allow it to return the rows in a hierarchical fashion where child rows are provided as a property on a row called `subRows`. For this to work as expected it is required that a flat list of data (that contains implicit hierarchy) be transformed into a hierarchical data structure.
+
+A third party library called [`performant-array-to-tree`](https://www.npmjs.com/package/performant-array-to-tree) offers an easy, and as the package says, performant means of doing this (in O(n) time). It comes with an MIT license, and is apparently fully tested, so it seems like it would be suitable to use for this purpose.
+
+### Group rows with data hierarchy
+
+Tanstack allows limited support of data modeling where there is both grouping and parent-child row relationships. Essentially, when there is grouping present, only rows at the top-level of a hierarchical set of data will be grouped. All further child rows under the top-level parent row will not be grouped.
```

Comment:
I'm wondering if we can clarify the last sentence of this paragraph. Maybe something like _Child rows will continue to be shown under their parent row rather than being grouped based on their own data._